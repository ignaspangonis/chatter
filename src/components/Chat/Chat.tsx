import React, { useEffect, useRef } from 'react'
import { Button } from 'react-daisyui'

import { Message } from 'src/types/models'

import SendMessage from './SendMessage/SendMessage'

type Props = {
  messages: Message[]
  onSendMessage: (message: string) => void
  onCloseConnection: () => void
}

export default function Chat({ messages, onSendMessage, onCloseConnection }: Props) {
  const messageRef = useRef<HTMLDivElement>(null)

  useEffect(() => {
    if (!messageRef.current) return

    const { scrollHeight, clientHeight } = messageRef.current

    // TODO: make this work
    messageRef.current.scrollTo({ left: 0, top: scrollHeight - clientHeight, behavior: 'smooth' })
  }, [messages])

  return (
    <div ref={messageRef} className="w-full">
      <div className="mb-regular flex justify-end">
        <Button onClick={onCloseConnection}>Leave Room</Button>
      </div>
      <div>
        <div className="overflow-auto min-h-[400px] rounded m-x-small bg-cg5">
          {messages.map(message => (
            <div className="user-message">
              <div className="text-sm">{message.userName}</div>
              <div className="inline-flex mb-[0] mt-small mx-auto p-small text-base text-cg7 rounded-md bg-primary">
                {message.message}
              </div>
            </div>
          ))}
        </div>
        <SendMessage onSubmit={onSendMessage} />
      </div>
    </div>
  )
}
