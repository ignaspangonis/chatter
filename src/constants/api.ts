export const ROOT_API_URL = 'https://localhost:7276'
