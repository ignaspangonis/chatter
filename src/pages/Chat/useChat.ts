import { useCallback, useContext, useEffect, useState } from 'react'

import { MessageModel } from 'src/libs/chat-room/types/models'
import { ChatContext } from 'src/containers/ChatProvider/ChatContext'

const handleError = (userMessage: string, error: unknown) => {
  alert(userMessage)

  if (error instanceof Error) {
    console.error(`${userMessage}. Error: ${error.message}. Stacktrace: ${error.stack}`)
  }
}

const useChat = (userName: string | null, roomName: string | null) => {
  const [messages, setMessages] = useState<MessageModel[]>([])
  const [users, setUsers] = useState<string[]>([])
  const { chatClient } = useContext(ChatContext)

  const handleConnectionClosed = useCallback(() => {
    setUsers([])
    setMessages([])
  }, [])

  const disconnectFromRoom = useCallback(async () => {
    if (!chatClient) return

    try {
      await chatClient.disconnect()
    } catch (error) {
      handleError('Failed to disconnect from the room, please try again later', error)
    }

    handleConnectionClosed()
  }, [handleConnectionClosed, chatClient])

  useEffect(() => {
    if (!userName || !roomName) return
    if (!chatClient) return

    chatClient.connect({
      roomName,
      userName,
      onGetUsers: setUsers,
      onNewMessage: message => setMessages(prevMessages => [...prevMessages, message]),
      onGetMessageHistory: messages => setMessages(messages),
      onClose: handleConnectionClosed,
      onError: error => handleError('Failed to connect, please try again later', error),
    })

    return () => {
      chatClient.disconnect().catch(error => console.error('Failed to disconnect', error))
    }
  }, [roomName, userName, chatClient, handleConnectionClosed])

  const sendMessage = useCallback(
    async (message: string) => {
      if (!chatClient) return

      try {
        await chatClient.sendMessage(message)
      } catch (error) {
        handleError('Failed to send message', error)
      }
    },
    [chatClient],
  )

  return {
    users,
    messages,
    disconnectFromRoom,
    sendMessage,
  }
}

export default useChat
