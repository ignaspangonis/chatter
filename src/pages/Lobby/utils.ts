type ValidateFormReturnType =
  | {
      isValid: true
    }
  | {
      isValid: false
      error: string
    }

const FORBIDDEN_USER_NAMES = ['admin', 'moderator', 'chatbot']

export const validateForm = (userName: string, roomName: string): ValidateFormReturnType => {
  if (!userName) {
    return { isValid: false, error: 'Please enter your username!' }
  }

  if (!roomName) {
    return { isValid: false, error: 'Please enter room name!' }
  }

  if (FORBIDDEN_USER_NAMES.includes(userName.toLowerCase())) {
    return { isValid: false, error: 'This username is not allowed! Please choose a different one.' }
  }

  return { isValid: true }
}
