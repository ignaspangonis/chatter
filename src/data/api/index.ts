import { ROOT_URL } from 'src/constants/api'
import { WeatherDto } from 'src/types/dtos'
import api from 'src/libs/api-client'

import * as response from './response'

export const getCurrentWeather = () => api.get<WeatherDto>(`${ROOT_URL}/weather`)
export const deleteMessageRoom = (roomName: string) =>
  api.delete<response.DeleteMessageRoomResponse>(`${ROOT_URL}/message-room/${roomName}`)
