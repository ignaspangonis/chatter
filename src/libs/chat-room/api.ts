import { ROOT_API_URL } from 'src/constants/api'
import api from 'src/libs/api-client'

type DeleteChatRoomResponse = {
  message: 'Room deleted'
}

export const deleteChatRoom = (roomName: string) =>
  api.delete<DeleteChatRoomResponse>(`${ROOT_API_URL}/message-room/${roomName}`)
