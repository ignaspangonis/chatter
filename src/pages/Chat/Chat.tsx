import React, { useEffect, useMemo, useRef } from 'react'
import { useNavigate, useSearchParams } from 'react-router-dom'

import { getLocalisedDate } from 'src/libs/utils/date'
import { Route } from 'src/constants/routes'
import { MessageModel } from 'src/libs/chat-room/types/models'

import SendMessage from './SendMessage'
import useChat from './useChat'
import AdminAction from './AdminAction'
import classNames from 'classnames'

export default function Chat() {
  const messageRef = useRef<HTMLDivElement>(null)

  const navigate = useNavigate()
  const [searchParams] = useSearchParams()

  const [userName, roomName] = useMemo(() => {
    const userName = searchParams.get('userName')
    const roomName = searchParams.get('roomName')

    if (!userName || !roomName) return [null, null]

    return [String(userName), String(roomName)]
  }, [searchParams])

  const { messages, users, disconnectFromRoom, sendMessage } = useChat(userName, roomName)

  useEffect(() => {
    if (!userName || !roomName) {
      alert('Connection was not found!')
      navigate(Route.Home)
    }
  }, [userName, roomName, navigate])

  useEffect(() => {
    if (!messageRef.current) return

    messageRef.current.scrollTo({
      top: messageRef.current.scrollHeight - messageRef.current.clientHeight,
      behavior: 'smooth',
      left: 0,
    })
  }, [messages])

  const renderMessage = (message: MessageModel, index: number) => {
    const isOwnMessage = message.userName === userName
    const isSystemMessage = message.userName === 'ChatBot'
    const isOtherUserMessage = !isOwnMessage && !isSystemMessage

    return (
      <div
        className={classNames('pr-small', {
          'text-right': isOwnMessage,
          'text-left': isOtherUserMessage,
          'text-center': isSystemMessage,
        })}
        role="gridcell"
        tabIndex={0}
        key={index}
      >
        {!isSystemMessage && <p className="text-sm">{message.userName}</p>}
        <p
          className={classNames(
            'inline-block mb-[0] mt-x-small mx-auto py-small px-regular text-base rounded-lg wrap-words z-large tooltip',
            {
              'text-cg2 text-xs tooltip-right': isSystemMessage,
              'bg-primary text-cg7 tooltip-left': isOwnMessage,
              'bg-base-300 text-cg1 tooltip-right': isOtherUserMessage,
            },
          )}
          data-tip={getLocalisedDate(message.createdAt)}
        >
          {message.content}
        </p>
      </div>
    )
  }

  function leaveRoom() {
    disconnectFromRoom()
    navigate(Route.Home)
  }

  return (
    <section className="w-full">
      <div className="mb-large flex justify-end items-center gap-large">
        <h2 className="text-2xl font-bold mr-auto">
          Room: <span className="text-accent">{roomName}</span>
        </h2>
        <AdminAction roomName={roomName} onBeforeDeleteRoom={leaveRoom} />
        <button className="btn btn-primary" onClick={leaveRoom}>
          Leave Room
        </button>
      </div>

      <div className="flex gap-large justify-between">
        <div className="mr-medium">
          <h3 className="text-lg font-bold">Connected users</h3>
          {users.map((user, index) => (
            <div key={index}>{user}</div>
          ))}
        </div>

        <div className="flex-1">
          <div
            className="h-[420px] overflow-y-auto flex flex-col gap-regular rounded-lg mb-large p-regular bg-base-200"
            ref={messageRef}
          >
            {messages.map(renderMessage)}
          </div>
          <SendMessage onSubmit={sendMessage} />
        </div>
      </div>
    </section>
  )
}
