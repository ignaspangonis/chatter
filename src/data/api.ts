import { ROOT_API_URL } from 'src/constants/api'
import { WeatherDto } from 'src/libs/chat-room/types'
import api from 'src/libs/api-client'

export const getCurrentWeather = () => api.get<WeatherDto>(`${ROOT_API_URL}/weather`)

type DeleteChatRoomResponse = {
  message: string
}

export const deleteChatRoom = (roomName: string) =>
  api.delete<DeleteChatRoomResponse>(`${ROOT_API_URL}/message-room/${roomName}`)
