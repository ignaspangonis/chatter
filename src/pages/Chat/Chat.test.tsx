import { act, render, screen, waitFor } from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import * as api from 'src/data/api'
import { MessageModel } from 'src/libs/chat-room/types'
import { ChatClientMock, withChat, withRouter, WithRouterOptions } from 'src/libs/utils/test'

import Chat from './Chat'

const buildMessageModel = (override?: Partial<MessageModel>): MessageModel => ({
  id: 'adfa',
  userName: 'John',
  content: 'Hello world!',
  createdAt: '2023-09-09T00:00:00.000Z',
  roomName: '1',
  ...override,
})

describe('<Chat />', () => {
  let routerOptions: WithRouterOptions
  let chatClient: ChatClientMock

  const deleteChatRoom = jest.spyOn(api, 'deleteChatRoom')

  const renderHelper = (component: JSX.Element) =>
    render(withChat(withRouter(component, routerOptions), chatClient))

  beforeEach(() => {
    deleteChatRoom.mockReturnValue(new Promise(() => {}))

    chatClient = new ChatClientMock()
    routerOptions = { initialEntries: ['?userName=John&roomName=best-room'] }
  })

  it('renders the room name', () => {
    renderHelper(<Chat />)

    expect(screen.getByText('best-room')).toBeInTheDocument()
  })

  it('renders messages with usernames', async () => {
    renderHelper(<Chat />)

    await waitFor(() => expect(chatClient.connect).toHaveBeenCalledTimes(1))
    expect(screen.queryByText('Hello world!')).not.toBeInTheDocument()

    const { onGetMessageHistory } = chatClient.connect.mock.calls[0][0]

    const messages = [
      buildMessageModel({ content: 'Hey world!', userName: 'Josh' }),
      buildMessageModel({ content: 'Hello!', userName: 'Mark' }),
    ]

    act(() => onGetMessageHistory(messages))

    expect(await screen.findByText('Hey world!')).toBeInTheDocument()
    expect(screen.getByText('Josh')).toBeInTheDocument()
    expect(screen.getByText('Hello!')).toBeInTheDocument()
    expect(screen.getByText('Mark')).toBeInTheDocument()
  })

  it('renders a new message with username', async () => {
    renderHelper(<Chat />)

    await waitFor(() => expect(chatClient.connect).toHaveBeenCalledTimes(1))
    expect(screen.queryByText('Hello world!')).not.toBeInTheDocument()
    expect(screen.queryByText('John')).not.toBeInTheDocument()

    const { onNewMessage } = chatClient.connect.mock.calls[0][0]
    const newMessage = buildMessageModel({ userName: 'John', content: 'Hello world!' })

    act(() => onNewMessage(newMessage))

    expect(await screen.findByText('Hello world!')).toBeInTheDocument()
    expect(screen.getByText('John')).toBeInTheDocument()
  })

  describe('when the user is admin', () => {
    beforeEach(() => {
      routerOptions = { initialEntries: ['?userName=John&roomName=1&admin=true'] }
    })

    it('calls api to delete room', async () => {
      renderHelper(<Chat />)

      userEvent.click(await screen.findByText('Delete room'))

      await waitFor(() => expect(deleteChatRoom).toHaveBeenCalledTimes(1))
    })
  })

  it('sends message', async () => {
    renderHelper(<Chat />)

    userEvent.type(screen.getByPlaceholderText('Aa'), 'Hello world!')
    userEvent.click(screen.getByText('Send'))

    await waitFor(() => expect(chatClient.sendMessage).toHaveBeenCalledTimes(1))
    expect(chatClient.sendMessage).toHaveBeenCalledWith('Hello world!')
  })

  it('displays user input', async () => {
    renderHelper(<Chat />)

    userEvent.type(screen.getByPlaceholderText('Aa'), 'Hello world!')

    expect(await screen.findByDisplayValue('Hello world!')).toBeInTheDocument()
  })
})
