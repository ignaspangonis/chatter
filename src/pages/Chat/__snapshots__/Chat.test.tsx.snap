// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chat /> renders correctly 1`] = `
<div>
  <section
    class="w-full"
  >
    <div
      class="mb-large flex justify-end items-center gap-large"
    >
      <h2
        class="text-2xl font-bold mr-auto"
      >
        Room: 
        <span
          class="text-accent"
        >
          Food
        </span>
      </h2>
      <button
        class="btn"
      >
        Make me admin
      </button>
      <button
        class="btn btn-primary"
      >
        Leave Room
      </button>
    </div>
    <div
      class="flex gap-large justify-between"
    >
      <div
        class="mr-medium"
      >
        <h3
          class="text-lg font-bold"
        >
          Connected users
        </h3>
        <div>
          Mark
        </div>
        <div>
          John
        </div>
      </div>
      <div
        class="flex-1"
      >
        <div
          class="h-[420px] overflow-y-auto flex flex-col gap-regular rounded-lg mb-large p-regular bg-cg6"
        >
          <div
            class="text-right pr-small"
            role="gridcell"
            tabindex="0"
          >
            <div
              class="text-sm"
            >
              Mark
            </div>
            <p
              class="inline-block text-left mb-[0] mt-x-small mx-auto py-small px-regular text-base text-cg7 rounded-lg bg-primary tooltip tooltip-left wrap-words"
              data-tip="8/1/2021, 3:00:00 AM"
            >
              Nice
            </p>
          </div>
        </div>
        <form
          class="SendMessage"
        >
          <div
            class="flex flex-row gap-regular grow"
          >
            <input
              class="input input-bordered mx-auto flex-grow"
              placeholder="Aa"
              value=""
            />
            <div
              class=""
            >
              <button
                class="btn"
                disabled=""
                type="submit"
              >
                Send
              </button>
            </div>
          </div>
        </form>
      </div>
    </div>
  </section>
</div>
`;
